// bertui/src/build/compiler/file-transpiler.js
// bertui/src/build/compiler/file-transpiler.js - FORCE PRODUCTION JSX
import { join, relative, dirname, extname } from 'path';
import { readdirSync, statSync, mkdirSync, writeFileSync } from 'fs';
import logger from '../../logger/logger.js';
import { replaceEnvInCode } from '../../utils/env.js';
import { transform } from 'lightningcss';

// ============================================
// CSS MODULES
// ============================================

function hashClassName(filename, className) {
  const str = filename + className;
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = (hash << 5) - hash + str.charCodeAt(i);
    hash |= 0;
  }
  return Math.abs(hash).toString(36).slice(0, 5);
}

function scopeCSSModule(cssText, filename) {
  const classNames = new Set();
  const classRegex = /\.([a-zA-Z_][a-zA-Z0-9_-]*)\s*[{,\s:]/g;
  let match;
  while ((match = classRegex.exec(cssText)) !== null) {
    classNames.add(match[1]);
  }
  const mapping = {};
  for (const cls of classNames) {
    mapping[cls] = `${cls}_${hashClassName(filename, cls)}`;
  }
  let scopedCSS = cssText;
  for (const [original, scoped] of Object.entries(mapping)) {
    scopedCSS = scopedCSS.replace(
      new RegExp(`\\.${original}(?=[\\s{,:\\[#.>+~)\\]])`, 'g'),
      `.${scoped}`
    );
  }
  return { mapping, scopedCSS };
}

async function compileCSSModule(srcPath, rootBuildDir) {
  const filename = srcPath.split('/').pop();
  const cssText = await Bun.file(srcPath).text();
  const { mapping, scopedCSS } = scopeCSSModule(cssText, filename);

  let finalCSS = scopedCSS;
  try {
    const { code } = transform({
      filename,
      code: Buffer.from(scopedCSS),
      minify: false,
      drafts: { nesting: true },
      targets: { chrome: 90 << 16 }
    });
    finalCSS = code.toString();
  } catch (e) {
    logger.warn(`LightningCSS failed for ${filename}: ${e.message}`);
  }

  // JS mapping → .bertuibuild/styles/home.module.css.js (for JS imports)
  const stylesDir = join(rootBuildDir, 'styles');
  mkdirSync(stylesDir, { recursive: true });
  const jsContent = `// CSS Module: ${filename} — auto-generated by BertUI\nconst styles = ${JSON.stringify(mapping, null, 2)};\nexport default styles;\n`;
  writeFileSync(join(stylesDir, filename + '.js'), jsContent);

  // Scoped CSS → .bertuibuild/styles-staged/home.module.css (for css-builder.js to pick up)
  const stagingDir = join(rootBuildDir, 'styles-staged');
  mkdirSync(stagingDir, { recursive: true });
  writeFileSync(join(stagingDir, filename), finalCSS);

  logger.debug(`CSS Module: ${filename} → ${Object.keys(mapping).length} classes scoped`);
}

function transformCSSModuleImports(code, outPath, rootBuildDir) {
  const moduleImportRegex = /import\s+(\w+)\s+from\s+['"]([^'"]*\.module\.css)['"]/g;
  const outFileDir = dirname(outPath);
  const stylesDir = join(rootBuildDir, 'styles');

  code = code.replace(moduleImportRegex, (match, varName, importPath) => {
    const filename = importPath.split('/').pop();
    const jsFile = join(stylesDir, filename + '.js');
    let rel = relative(outFileDir, jsFile).replace(/\\/g, '/');
    if (!rel.startsWith('.')) rel = './' + rel;
    return `import ${varName} from '${rel}'`;
  });

  return code;
}

// ============================================
// MAIN COMPILER
// ============================================

export async function compileBuildDirectory(srcDir, buildDir, root, envVars) {
  const bunfigContent = `
[build]
jsx = "react"
jsxFactory = "React.createElement"
jsxFragment = "React.Fragment"
`.trim();

  writeFileSync(join(buildDir, 'bunfig.toml'), bunfigContent);
  logger.info('Created bunfig.toml for classic JSX');

  const files = readdirSync(srcDir);
  const filesToCompile = [];
  const rootBuildDir = join(root, '.bertuibuild');

  for (const file of files) {
    const srcPath = join(srcDir, file);
    const stat = statSync(srcPath);

    if (stat.isDirectory()) {
      const subBuildDir = join(buildDir, file);
      mkdirSync(subBuildDir, { recursive: true });
      await compileBuildDirectory(srcPath, subBuildDir, root, envVars);
    } else {
      const ext = extname(file);

      if (file.endsWith('.module.css')) {
        await compileCSSModule(srcPath, rootBuildDir);
        continue;
      }

      if (ext === '.css') continue;

      if (['.jsx', '.tsx', '.ts'].includes(ext)) {
        filesToCompile.push({ path: srcPath, dir: buildDir, name: file, type: 'tsx' });
      } else if (ext === '.js') {
        filesToCompile.push({ path: srcPath, dir: buildDir, name: file, type: 'js' });
      }
    }
  }

  if (filesToCompile.length === 0) return;

  logger.info(`📦 Compiling ${filesToCompile.length} files...`);

  for (let i = 0; i < filesToCompile.length; i++) {
    const file = filesToCompile[i];

    try {
      if (file.type === 'tsx') {
        await compileBuildFile(file.path, file.dir, file.name, root, envVars, rootBuildDir);
      } else {
        await compileJSFile(file.path, file.dir, file.name, root, envVars, rootBuildDir);
      }

      if ((i + 1) % 10 === 0 || i === filesToCompile.length - 1) {
        const percent = (((i + 1) / filesToCompile.length) * 100).toFixed(0);
        logger.info(`   Progress: ${i + 1}/${filesToCompile.length} (${percent}%)`);
      }

    } catch (error) {
      logger.error(`Failed to compile ${file.name}: ${error.message}`);
    }
  }

  logger.success(`✅ Compiled ${filesToCompile.length} files`);
}

async function compileBuildFile(srcPath, buildDir, filename, root, envVars, rootBuildDir) {
  const ext = extname(filename);

  try {
    let code = await Bun.file(srcPath).text();
    code = replaceEnvInCode(code, envVars);

    const outFilename = filename.replace(/\.(jsx|tsx|ts)$/, '.js');
    const outPath = join(buildDir, outFilename);

    code = transformCSSModuleImports(code, outPath, rootBuildDir);
    code = removePlainCSSImports(code);
    code = fixBuildImports(code, srcPath, outPath, root);

    if (!code.includes('import React')) {
      code = `import React from 'react';\n${code}`;
    }

    const transpiler = new Bun.Transpiler({
      loader: ext === '.tsx' ? 'tsx' : ext === '.ts' ? 'ts' : 'jsx',
      target: 'browser',
      define: { 'process.env.NODE_ENV': '"production"' },
      tsconfig: {
        compilerOptions: {
          jsx: 'react',
          jsxFactory: 'React.createElement',
          jsxFragmentFactory: 'React.Fragment',
          target: 'ES2020'
        }
      }
    });

    let compiled = await transpiler.transform(code);

    if (compiled.includes('jsxDEV')) {
      logger.warn(`⚠️  Dev JSX detected in ${filename}, fixing...`);
      compiled = compiled.replace(/jsxDEV/g, 'jsx');
    }

    compiled = fixRelativeImports(compiled);
    await Bun.write(outPath, compiled);

    code = null;
    compiled = null;

  } catch (error) {
    logger.error(`Failed to compile ${filename}: ${error.message}`);
    throw error;
  }
}

async function compileJSFile(srcPath, buildDir, filename, root, envVars, rootBuildDir) {
  const outPath = join(buildDir, filename);
  let code = await Bun.file(srcPath).text();
  code = replaceEnvInCode(code, envVars);
  code = transformCSSModuleImports(code, outPath, rootBuildDir);
  code = removePlainCSSImports(code);
  code = fixBuildImports(code, srcPath, outPath, root);

  if (usesJSX(code) && !code.includes('import React')) {
    code = `import React from 'react';\n${code}`;
  }

  await Bun.write(outPath, code);
  code = null;
}

function usesJSX(code) {
  return code.includes('React.createElement') ||
         code.includes('React.Fragment') ||
         /<[A-Z]/.test(code);
}

function removePlainCSSImports(code) {
  code = code.replace(/import\s+['"][^'"]*(?<!\.module)\.css['"];?\s*/g, '');
  code = code.replace(/import\s+['"]bertui\/styles['"]\s*;?\s*/g, '');
  return code;
}

function fixBuildImports(code, srcPath, outPath, root) {
  const buildDir = join(root, '.bertuibuild');
  const routerPath = join(buildDir, 'router.js');
  const relativeToRouter = relative(dirname(outPath), routerPath).replace(/\\/g, '/');
  const routerImport = relativeToRouter.startsWith('.') ? relativeToRouter : './' + relativeToRouter;
  code = code.replace(/from\s+['"]bertui\/router['"]/g, `from '${routerImport}'`);
  return code;
}

function fixRelativeImports(code) {
  const importRegex = /from\s+['"](\.\.[\/\\]|\.\/)((?:[^'"]+?)(?<!\.js|\.jsx|\.ts|\.tsx|\.json))['"];?/g;
  code = code.replace(importRegex, (match, prefix, path) => {
    if (path.endsWith('/') || /\.\w+$/.test(path)) return match;
    return `from '${prefix}${path}.js';`;
  });
  return code;
}
